### ISSUE TITLE
Bounding Boxes Not Rotating with Image During YOLO11 Augmentation

### ISSUE THREAD

**Issue by yu04ko21no** [ultralytics/ultralytics#21472]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

"I’m new to training YOLO and I don’t fully understand how it works yet." 
I'm Japanese so my english is not good 

When I perform data augmentation with rotation, the image is rotated correctly, but it seems that the labeled bounding boxes are not rotated accordingly. This causes a mismatch between the object and the annotation in the training image.

How can I make sure that the labels rotate together with the image during augmentation?

I am using the official Ultralytics YOLO11 implementation.

I noticed this issue by checking the val_batch_labels.jpg file.

Is there a setting or method to apply the same geometric transformation (like rotation) to both image and label simultaneously?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @yu04ko21no, thank you for reaching out and for your interest in Ultralytics 🚀! Your English is clear—no worries! Please note this is an automated response to help you get started, and an Ultralytics engineer will also assist you soon.

We recommend visiting the [Docs](https://docs.ultralytics.com/) for helpful [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many common questions may already be answered there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) to help us investigate. This usually includes:
1. The exact augmentation settings you used
2. Relevant config or training command
3. Example images and labels before/after augmentation (if possible)

For training ❓ questions, please include as much detail as you can, such as dataset examples and training logs, and check you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join our community for more support and discussion:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time chat
- [Discourse](https://community.ultralytics.com/) for in-depth discussions
- [Subreddit](https://www.reddit.com/r/Ultralytics/) to share with others

## Upgrade

Please upgrade to the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment using [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in the following up-to-date, verified environments (with dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM ([GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/))
- **Amazon** Deep Learning AMI ([AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/))
- **Docker Image** ([Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks again for your report! An Ultralytics engineer will assist you soon.

**Comment by Y-T-G** [comment]
The boxes wouldn't be rotated unless you're training an OBB model. The boxes are always horizontal. They will be adjusted to fit the rotated object, but not rotated themselves.

**Comment by yu04ko21no** [comment]
> The boxes wouldn't be rotated unless you're training an OBB model. The boxes are always horizontal. They will be adjusted to fit the rotated object, but not rotated themselves.

.yaml ↓
hsv_h: 0.0
hsv_s: 0.0
hsv_v: 0.0
translate: 0.0
scale: 0.0
fliplr: 0.0
flipud: 0.0
degrees: 0
shear: 0
perspective: 0
mixup: 0.0
cutmix: 0.0
copy_paste: 0.0

The yaml file should be above, but the image is rotated during training

**Comment by SebastianJanampa** [comment]
you could try this

```python
from ultralytics import YOLO

# Load a model
model = YOLO("yolo11n.yaml")  # build a new model from YAML

# Train the model
results = model.train(data="coco8.yaml", epochs=100, imgsz=640, hsv_h=0, hsv_s=0, hsv_v=0, translate=0, scale=0, fliplr=0, flipud=0, degrees=0, shear=0, perspective=0, mixup=0, cutmix=0, copy_paste=0)

```

**Comment by yu04ko21no** [comment]
> you could try this
> 
> from ultralytics import YOLO
> 
> # Load a model
> model = YOLO("yolo11n.yaml")  # build a new model from YAML
> 
> # Train the model
> results = model.train(data="coco8.yaml", epochs=100, imgsz=640, hsv_h=0, hsv_s=0, hsv_v=0, translate=0, scale=0, fliplr=0, flipud=0, degrees=0, shear=0, perspective=0, mixup=0, cutmix=0, copy_paste=0)


What is the difference between "creating a new model myself with a yaml file" and "using YOLO11x.pt"?

**Comment by glenn-jocher** [comment]
The difference is in the starting weights:

- `YOLO("yolo11n.yaml")` builds a new model architecture from scratch with random weights - you're starting training from zero
- `YOLO("yolo11n.pt")` loads a pretrained model with weights already trained on datasets like COCO - this gives you a head start with transfer learning

For your augmentation issue, the model initialization method doesn't matter. If images are still rotating despite setting `degrees=0`, there might be augmentations happening elsewhere in your pipeline or dataset preparation. Can you share how you're loading your dataset and the exact training command you're using?

**Comment by yu04ko21no** [comment]
> The difference is in the starting weights:
> 
> * `YOLO("yolo11n.yaml")` builds a new model architecture from scratch with random weights - you're starting training from zero
> * `YOLO("yolo11n.pt")` loads a pretrained model with weights already trained on datasets like COCO - this gives you a head start with transfer learning
> 
> For your augmentation issue, the model initialization method doesn't matter. If images are still rotating despite setting `degrees=0`, there might be augmentations happening elsewhere in your pipeline or dataset preparation. Can you share how you're loading your dataset and the exact training command you're using?

### I want to create a model that can detect 5 different types of LEGO blocks.

**train.py**

from ultralytics import YOLO 
model = YOLO("yolo11x.pt")  # build a new model from YAML

results = model.train(data= "C:\YOLO\data.yaml" , epochs=100, imgsz=640, hsv_h=0, hsv_s=0, hsv_v=0, translate=0, scale=0, fliplr=0, flipud=0, degrees=0, shear=0, perspective=0, mixup=0, cutmix=0, copy_paste=0)


**data.yaml**

train : images\train
val : images\val
nc : 5
names : ['R_13' , 'R_14' , 'R_22' , 'Y_14' , 'Y_23']   #〇_〇〇 means "color" and "Height x width size"


![Image](https://github.com/user-attachments/assets/806eea51-3a6b-40d2-abe3-f6903de15b85)

**Comment by SebastianJanampa** [comment]
@yu04ko21no 
Try to use augmentations, otherwise, your model might overfit and will not be robust. I would just keep the ` hsv_h=0, hsv_s=0, hsv_v=0` since it looks like R_14 and Y_14 are the same structure but with different colors.

**Comment by yu04ko21no** [comment]
<img width="681" height="696" alt="Image" src="https://github.com/user-attachments/assets/48ccdd19-e295-47d7-966a-337b9756549a" />

In the runtime program, each parameter of the augmentation is set to 0, but this image doesn't seem to be working.

**Comment by SebastianJanampa** [comment]
Is your dataset in the format `x1,y1,x2,y2`?

**Comment by yu04ko21no** [comment]
> Is your dataset in the format `x1,y1,x2,y2`?

Coordinates indicating where the annotated object is located. Normalized

Label number, left end, right end, top end, bottom end of object 
4 0.317155 0.785482 0.138354 0.226190

**Comment by SebastianJanampa** [comment]
@yu04ko21no sorry for the mistake.
To train using ultralytics, the format of your dataset must be `xywh` (normalized) `xy` are the the center of the bounding box

**Comment by yu04ko21no** [comment]
> [@yu04ko21no](https://github.com/yu04ko21no) sorry for the mistake. To train using ultralytics, the format of your dataset must be `xywh` (normalized) `xy` are the the center of the bounding box

ok
i mistake too
However, I think the coordinates themselves are the correct ones, since I put together a script that automatically converts them to YOLO format before returning them.

The problem is that even with the extension turned off, cutmix and other operations still happen on their own.

**Comment by glenn-jocher** [comment]
You're missing the `mosaic` parameter - it defaults to 1.0 and combines four training images into one, which would explain the transformations you're seeing. Add `mosaic=0.0` to your training command:

```python
results = model.train(data="C:\YOLO\data.yaml", epochs=100, imgsz=640, hsv_h=0, hsv_s=0, hsv_v=0, translate=0, scale=0, fliplr=0, flipud=0, degrees=0, shear=0, perspective=0, mixup=0, cutmix=0, copy_paste=0, mosaic=0.0)
```

The [mosaic augmentation](https://docs.ultralytics.com/modes/train/#augmentation-settings-and-hyperparameters) is enabled by default and creates composite images from multiple training samples, which could be what you're observing in your validation batch images.

**Comment by yu04ko21no** [comment]
> You're missing the `mosaic` parameter - it defaults to 1.0 and combines four training images into one, which would explain the transformations you're seeing. Add `mosaic=0.0` to your training command:
> 
> results = model.train(data="C:\YOLO\data.yaml", epochs=100, imgsz=640, hsv_h=0, hsv_s=0, hsv_v=0, translate=0, scale=0, fliplr=0, flipud=0, degrees=0, shear=0, perspective=0, mixup=0, cutmix=0, copy_paste=0, mosaic=0.0)
> The [mosaic augmentation](https://docs.ultralytics.com/modes/train/#augmentation-settings-and-hyperparameters) is enabled by default and creates composite images from multiple training samples, which could be what you're observing in your validation batch images.



I see!
I'll give it a try. 
By the way, can't we use these parameters to change the size and location of the photo in object detection? What are other people doing?
I feel like it would be fatal if the bounding box is immutable.

**Comment by glenn-jocher** [comment]
You're misunderstanding how it works - the bounding boxes ARE transformed during geometric augmentations, they just remain axis-aligned (horizontal rectangles). When YOLO applies rotation, translation, or scaling, the [RandomPerspective class](https://docs.ultralytics.com/reference/data/augment/#ultralytics.data.augment.RandomPerspective) automatically adjusts the bounding boxes to fit the transformed objects using methods like `apply_bboxes()` and `apply_segments()`. The boxes expand or move as needed to encompass the rotated/scaled objects, but stay as standard horizontal rectangles rather than becoming rotated rectangles themselves. This is exactly what most people use for training robust object detection models - the geometric augmentations help the model generalize to different orientations and scales while keeping the annotation format consistent.

**Comment by yu04ko21no** [comment]
> You're misunderstanding how it works - the bounding boxes ARE transformed during geometric augmentations, they just remain axis-aligned (horizontal rectangles). When YOLO applies rotation, translation, or scaling, the [RandomPerspective class](https://docs.ultralytics.com/reference/data/augment/#ultralytics.data.augment.RandomPerspective) automatically adjusts the bounding boxes to fit the transformed objects using methods like `apply_bboxes()` and `apply_segments()`. The boxes expand or move as needed to encompass the rotated/scaled objects, but stay as standard horizontal rectangles rather than becoming rotated rectangles themselves. This is exactly what most people use for training robust object detection models - the geometric augmentations help the model generalize to different orientations and scales while keeping the annotation format consistent.



I understand that it is a horizontal rectangle.
The problem is that when various extensions are made, the position of the original bounding box is so strange that learning is not possible.

I understand that extensions are essential for generalization in object detection. That's why I'm curious if others are experiencing the same situation as me when making extensions. And I don't know how everyone else is dealing with it.

I'm just annotating images, labeling them, splitting them into train and validation sets, and training with YOLO...

**Comment by glenn-jocher** [comment]
It sounds like there may be an issue with how the transformations are being applied to your specific dataset. To debug this, can you check your `val_batch0_labels.jpg` file with mosaic disabled and compare it to the original annotated images? 

Also verify your label format is correct - YOLO expects `class x_center y_center width height` (all normalized 0-1), not `class x1 y1 x2 y2`. If your conversion script has any issues, that could cause the misalignment you're seeing during augmentation.

If the boxes still look incorrect after disabling mosaic, please share a specific example showing the original image + labels vs the augmented result so we can identify what's going wrong.

**Comment by yu04ko21no** [comment]
> It sounds like there may be an issue with how the transformations are being applied to your specific dataset. To debug this, can you check your `val_batch0_labels.jpg` file with mosaic disabled and compare it to the original annotated images?
> 
> Also verify your label format is correct - YOLO expects `class x_center y_center width height` (all normalized 0-1), not `class x1 y1 x2 y2`. If your conversion script has any issues, that could cause the misalignment you're seeing during augmentation.
> 
> If the boxes still look incorrect after disabling mosaic, please share a specific example showing the original image + labels vs the augmented result so we can identify what's going wrong.



<img width="2400" height="1200" alt="Image" src="https://github.com/user-attachments/assets/8b40aa60-4c53-4a23-b394-325b6a74f823" />
![Image](https://github.com/user-attachments/assets/b399ff44-ff0b-46b7-b40f-378277debab4)


My basic premise is that I am a beginner, so I don't know much about it. Feel free to point this out. It will help me learn.

The attached image is val_batch0labels.jpg and the results of the study.
The numbers in the annotated text file seem to be correct, as the annotated frames are well reflected in some images.



Part of the train.py code.

# Train the model 
results = model.train(data= "C:\YOLO\data.yaml" , epochs=100, imgsz=640, device=0 , hsv_h=0, hsv_s=0, hsv_v=0, translate=0, scale =0, fliplr=0, flipud=0, degrees=0, shear=0, perspective=0, mixup=0, cutmix=0, copy_paste=0)



I forgot to turn off “mosaic” for sure, so I will try again.

On the other hand, why do I have to set so many parameters to 0 to get it to work, when extensions are supposed to be so important in deep learning?

In my present case, the background is fixed and the data expansion may not be that important, but I was wondering a bit about the general

Translated with [DeepL.com](https://www.deepl.com/?utm_campaign=product&utm_source=web_translator&utm_medium=web&utm_content=copy_free_translation) (free version)

**Comment by yu04ko21no** [comment]
![Image](https://github.com/user-attachments/assets/8a1ec367-cb11-4d82-a597-29a04ef2dfcb)

val_batch0_labels.jpg

**Comment by SebastianJanampa** [comment]
Hi @yu04ko21no ,

I can notice that the classification works well most of the time, but the real problem here is the estimation of the bounding boxes. 
Images 156, 152, 138, and 151 appear to be distorted. Something useful would be to manually inspect your dataset and check that all the ground-truths are correct, and that you  have exported them in the correct format.

Now, about the augmentations. We utilize augmentation techniques to enhance the variety of images our model learns from, without increasing the number of images we store. Why do we do this? To avoid overfitting and to improve generalization. There's no exact recipe for which augmentation we must use. Everything will depend on the problem we are trying to solve. For example, you could use the `RandomCropping` technique, which randomly chooses a part of the image. However, in your case, that might be dangerous since only a small part of your image contains the object you want to detect. Another example is the augmentations that change the color of the image. For your specific case, adding that augmentation would hurt your predictions instead of improving, since your classification depends on the color of the object. Augmentations you could use are `flipping` and `geometric transformations`

**Comment by yu04ko21no** [comment]
To everyone who commented here,
I was unable to successfully convert the annotations to a txt file.

I annotated a vertical image, but it was saved in a horizontal format after being rotated 90 degrees.
Horizontal images were saved as horizontal.

Since the annotations were not done correctly, the learning process was naturally unsuccessful.

I am truly sorry.
I will try again on my own.
Thank you very much.

Translated with [DeepL.com](https://www.deepl.com/?utm_campaign=product&utm_source=web_translator&utm_medium=web&utm_content=copy_free_translation) (free version)

**Comment by yu04ko21no** [comment]
To everyone who commented here, I was unable to successfully convert the annotations to a txt file.

I annotated a vertical image, but it was saved in a horizontal format after being rotated 90 degrees. Horizontal images were saved as horizontal.

Since the annotations were not done correctly, the learning process was naturally unsuccessful.

 I am truly sorry. I will try again on my own. Thank you very much.

**Comment by yu04ko21no** [comment]
And if the annotation is correct, various geometric transformations will probably be useful, so I will try them out.

**Comment by glenn-jocher** [comment]
Great detective work finding the root cause! Image orientation issues during annotation export are a common source of confusion. Once you have the annotations properly aligned with your images, the geometric augmentations will indeed be very helpful for training a robust LEGO detection model.

For future reference, you can use the `visualize_image_annotations()` function from `ultralytics.data.utils` to quickly verify that your bounding boxes align correctly with objects before training. Good luck with your LEGO detector!
